File: android/src/main/java/com/securitysuite/JwsFetchSigner.java
package com.securitysuite;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;

import javax.crypto.SecretKey;

import java.nio.charset.StandardCharsets;

/**
 * Resolves JWS signing for {@link Sslpinning#fetch}.
 *
 * <p>Priority:
 * <ol>
 *   <li>{@code options.jws.secret} — RFC 7515 compact JWS with explicit secret string</li>
 *   <li>{@code options.jws} + native HMAC key — legacy detached JWS with {@code jws.headers}</li>
 *   <li>{@code options.keyId} + {@code options.requestId} + native HMAC key — legacy shortcut
 *       (signs raw request body; header {@code X-JWS-Signature})</li>
 *   <li>{@code options.keyId} + {@code options.requestId} + {@code options.secret} — explicit
 *       secret fallback for the legacy detached format</li>
 * </ol>
 *
 * <p>After {@code establishSharedKey} / {@code getSharedKey}, callers can pass only
 * {@code keyId} and {@code requestId} — no secret in JavaScript.
 */
public final class JwsFetchSigner {
  public static final String LEGACY_JWS_HEADER = "X-JWS-Signature";
  public static final String MODERN_JWS_HEADER = "X-Request-Signature";

  public static final class Result {
    public final String headerName;
    public final String signature;

    public Result(String headerName, String signature) {
      this.headerName = headerName;
      this.signature = signature;
    }
  }

  private JwsFetchSigner() {}

  public static Result sign(
      String url,
      String method,
      byte[] requestBody,
      ReadableMap options,
      SecretKey nativeHmacKey
  ) throws Exception {
    if (options.hasKey("jws") && options.getMap("jws") != null) {
      return signWithJwsOptions(url, method, requestBody, options.getMap("jws"), nativeHmacKey);
    }

    if (options.hasKey("keyId") && options.hasKey("requestId")) {
      return signWithLegacyKeyId(
          requestBody,
          options.getString("keyId"),
          options.getString("requestId"),
          options.hasKey("secret") ? options.getString("secret") : null,
          nativeHmacKey
      );
    }

    return null;
  }

  private static Result signWithJwsOptions(
      String url,
      String method,
      byte[] requestBody,
      ReadableMap jwsOptions,
      SecretKey nativeHmacKey
  ) throws Exception {
    String headerName = MODERN_JWS_HEADER;
    if (jwsOptions.hasKey("headerName") && jwsOptions.getString("headerName") != null) {
      headerName = jwsOptions.getString("headerName");
    }

    if (jwsOptions.hasKey("secret")) {
      String secret = jwsOptions.getString("secret");
      if (secret == null || secret.trim().isEmpty()) {
        throw new IllegalArgumentException("JWS secret is required and must be a non-empty string");
      }

      String payloadString = JwsFetchPayload.build(url, method, requestBody, jwsOptions);
      boolean detached = jwsOptions.hasKey("detached") && jwsOptions.getBoolean("detached");
      String algorithm = jwsOptions.hasKey("algorithm") ? jwsOptions.getString("algorithm") : null;
      ReadableMap headers = jwsOptions.hasKey("headers") ? jwsOptions.getMap("headers") : null;

      JWSGenerator generator = new JWSGenerator();
      String signature = generator.generate(
          payloadString,
          secret,
          algorithm,
          headers,
          detached
      );
      return new Result(headerName, signature);
    }

    if (nativeHmacKey == null) {
      throw new IllegalStateException(
          "HMAC key required for JWS signing. Call establishSharedKey or getSharedKey first."
      );
    }

    if (!jwsOptions.hasKey("headers") || jwsOptions.getMap("headers") == null) {
      throw new IllegalArgumentException("JWS headers (kid, request_id) are required");
    }

    JWSGenerator generator = new JWSGenerator();
    String signature = generator.generate(
        requestBody,
        nativeHmacKey,
        jwsOptions.hasKey("algorithm") ? jwsOptions.getString("algorithm") : "HS256",
        jwsOptions.getMap("headers")
    );
    return new Result(LEGACY_JWS_HEADER, signature);
  }

  private static Result signWithLegacyKeyId(
      byte[] requestBody,
      String keyId,
      String requestId,
      String explicitSecret,
      SecretKey nativeHmacKey
  ) throws Exception {
    byte[] payload = requestBody != null ? requestBody : new byte[0];
    JWSGenerator generator = new JWSGenerator();

    if (nativeHmacKey != null) {
      return new Result(
          LEGACY_JWS_HEADER,
          generator.jwsHeader(payload, keyId, requestId, nativeHmacKey)
      );
    }

    if (explicitSecret != null && !explicitSecret.trim().isEmpty()) {
      WritableMap legacyHeaders = Arguments.createMap();
      legacyHeaders.putString("kid", keyId);
      legacyHeaders.putString("request_id", requestId);
      String signature = generator.generate(
          new String(payload, StandardCharsets.UTF_8),
          explicitSecret,
          "HS256",
          legacyHeaders,
          true
      );
      return new Result(LEGACY_JWS_HEADER, signature);
    }

    throw new IllegalStateException(
        "HMAC key required for JWS signing. Call establishSharedKey or getSharedKey first."
    );
  }
}

File: ios/JwsFetchSigner.swift
import Foundation

/// Resolves JWS signing for `SecuritySuite.fetch`.
///
/// Priority:
/// 1. `options.jws.secret` — RFC 7515 compact JWS with an explicit secret string
/// 2. `options.jws` + native HMAC key — legacy detached JWS with `jws.headers`
/// 3. `options.keyId` + `options.requestId` + native HMAC key — legacy shortcut
///    (signs raw request body; header `X-JWS-Signature`)
/// 4. `options.keyId` + `options.requestId` + `options.secret` — explicit secret fallback
///
/// After `establishSharedKey` / `getSharedKey`, callers can pass only `keyId` and
/// `requestId` — no secret in JavaScript.
@available(iOS 13.0, *)
enum JwsFetchSigner {
    static let legacyHeaderName = "X-JWS-Signature"
    static let modernHeaderName = "X-Request-Signature"

    struct Result {
        let headerName: String
        let signature: String
    }

    static func sign(
        url: String,
        method: String,
        requestBody: Data?,
        options: NSDictionary,
        nativeHmacKey: Data?
    ) throws -> Result? {
        if let jwsOptions = options["jws"] as? [String: Any] {
            return try signWithJwsOptions(
                url: url,
                method: method,
                requestBody: requestBody,
                jwsOptions: jwsOptions,
                nativeHmacKey: nativeHmacKey
            )
        }

        if options["keyId"] != nil && options["requestId"] != nil {
            return try signWithLegacyKeyId(
                requestBody: requestBody,
                keyId: options["keyId"] as? String,
                requestId: options["requestId"] as? String,
                explicitSecret: options["secret"] as? String,
                nativeHmacKey: nativeHmacKey
            )
        }

        return nil
    }

    private static func signWithJwsOptions(
        url: String,
        method: String,
        requestBody: Data?,
        jwsOptions: [String: Any],
        nativeHmacKey: Data?
    ) throws -> Result {
        var headerName = modernHeaderName
        if let customHeaderName = jwsOptions["headerName"] as? String, !customHeaderName.isEmpty {
            headerName = customHeaderName
        }

        if let secret = jwsOptions["secret"] as? String,
           !secret.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            let headers = jwsOptions["headers"] as? [String: Any]
            let algorithm = jwsOptions["algorithm"] as? String
            let detached = jwsOptions["detached"] as? Bool ?? false
            let payloadString = try JwsFetchPayload.build(
                url: url,
                method: method,
                requestBody: requestBody,
                jwsOptions: jwsOptions
            )
            let signature = try JWSGenerator.generate(
                payloadString: payloadString,
                secret: secret,
                algorithm: algorithm,
                headers: headers,
                detached: detached
            )
            return Result(headerName: headerName, signature: signature)
        }

        guard let nativeHmacKey else {
            throw NSError(
                domain: "JWS",
                code: 20,
                userInfo: [
                    NSLocalizedDescriptionKey:
                        "HMAC key required for JWS signing. Call establishSharedKey or getSharedKey first.",
                ]
            )
        }

        let headers = jwsOptions["headers"] as? [String: String] ?? [:]
        let algorithm = jwsOptions["algorithm"] as? String
        let signature = try JWSGenerator.generate(
            payload: requestBody ?? Data(),
            secret: nativeHmacKey,
            algorithm: algorithm,
            headers: headers
        )
        return Result(headerName: legacyHeaderName, signature: signature)
    }

    private static func signWithLegacyKeyId(
        requestBody: Data?,
        keyId: String?,
        requestId: String?,
        explicitSecret: String?,
        nativeHmacKey: Data?
    ) throws -> Result {
        guard let keyId, let requestId else {
            throw NSError(
                domain: "JWS",
                code: 21,
                userInfo: [NSLocalizedDescriptionKey: "keyId and requestId are required for legacy JWS signing"]
            )
        }

        let payload = requestBody ?? Data()

        if let nativeHmacKey {
            let signature = try JWSGenerator.generate(
                payload: payload,
                secret: nativeHmacKey,
                algorithm: "HS256",
                headers: ["kid": keyId, "request_id": requestId]
            )
            return Result(headerName: legacyHeaderName, signature: signature)
        }

        if let explicitSecret,
           !explicitSecret.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            let payloadString = String(data: payload, encoding: .utf8) ?? ""
            let signature = try JWSGenerator.generate(
                payloadString: payloadString,
                secret: explicitSecret,
                algorithm: "HS256",
                headers: ["kid": keyId, "request_id": requestId],
                detached: true
            )
            return Result(headerName: legacyHeaderName, signature: signature)
        }

        throw NSError(
            domain: "JWS",
            code: 20,
            userInfo: [
                NSLocalizedDescriptionKey:
                    "HMAC key required for JWS signing. Call establishSharedKey or getSharedKey first.",
            ]
        )
    }
}
